src/lualib/Promise.ts

```diff
@@ -0,0 +1,335 @@
+    onrejected: ((reason?: any) => void) ) => void ) {
+    if ( executor ) {
+      try {
+        executor(
+          value => {
+            this.fulfill( value as T);
+          },
+          reason => {
+            this.reject( reason );
+          });
+      } catch ( err ) {
+        this.reject( err );
+      }
+    }
+  }
+
+  public then<TResult1 = T, TResult2 = never>(
+    onFulfilled?: ((value: T) => TResult1 | PromiseImpl<TResult1>) | undefined,
+    onRejected?: ((reason: any) => TResult2 | PromiseImpl<TResult2>) | undefined
+  ): PromiseImpl<TResult1 | TResult2> {
+    const promise = new PromiseImpl<TResult1 | TResult2>();
+
+    if ( this.state === PromiseState.PENDING ) {
+        this.queue.push( {
+          promise,
+          onFulfilled,
+          onRejected,
+        } as Deferred< T, TResult1 | TResult2 > );
+    } else if ( this.state === PromiseState.FULFILLED ) {
+        if ( onFulfilled ) {
+          runAtNextTick( () => {
+            try {
+              PromiseImpl.PromiseResolutionProcedure( promise, onFulfilled(this.value) as TResult1 );
+            } catch ( err ) {
+              PromiseImpl.reject( err );
+            }
+          });
+        } else {
+          return this as any;
+        }
+    }  else if ( this.state === PromiseState.REJECTED ) {
+        this.handledError = true;
+        if ( onRejected ) {
+          runAtNextTick( () => {
+            try {
+              PromiseImpl.PromiseResolutionProcedure( promise, onRejected(this.reason) as TResult2);
+            } catch ( err ) {
+              PromiseImpl.reject( err );
+            }
+          });
+        } else {
+          return this as any;
+        }
+    }
+
+    return promise;
+  }
+
+  public catch<TResult = never>(
+    onRejected?: ((reason: any) => TResult | PromiseImpl<TResult>) | undefined
+  ): PromiseImpl<T | TResult> {
+    return this.then( undefined, onRejected );
+  }
+
+  protected resolve( value: T ): void {
+    PromiseImpl.PromiseResolutionProcedure<T>( this, value );
+  }
+
+  protected reject( reason: any ): void {
+    if ( this.state !== PromiseState.PENDING ) {
+      return;
+    }
+    this.state = PromiseState.REJECTED;
+    this.reason = reason;
+    this.queue.forEach( d => {
+      if ( typeof d.onRejected !== "function" ) { // if reject isnt a function -> ignore it
+        d.promise.reject(this.reason);
+      } else {
+        runAtNextTick( () => {
+          try {
+            const val = (d.onRejected as any)( reason );
+            PromiseImpl.PromiseResolutionProcedure( d.promise, val );
+          } catch ( err ) {
+            d.promise.reject( err );
+          }
+        });
+      }
+    });
+  }
+
+  protected __gc(): void {
+    if ( this.state === PromiseState.REJECTED && !this.handledError ) {
+      for ( const d of this.queue ) {
+        if ( d.onRejected ) {
+          return;
+        }
+      }
+      logError("Unhandled rejected promise ", this.reason );
```

Comment:
Might be useful to provide something like `unhandledrejection` event in js

Reply:
Since there is not window Object with event handling etc I chose an approach similar to the _setAtNextTick( ) function:

```typescript
interface Promise {
  _setUnhandledRejectionCallback( func: ( reason: string ) => void ): void;
}
Promise._setUnhandledRejectionCallback( ( reason ) => print(reason) );
```